huxley/core/models.py

```diff
@@ -444,3 +270,195 @@ def save(self, *args, **kwargs):
     class Meta:
         db_table = u'delegate'
         ordering = ['school']
+
+
+class Registration(models.Model):
+    CONTACT_TYPE_OPTIONS = ((ContactType.FACULTY, 'Faculty'),
+                            (ContactType.STUDENT, 'Student'), )
+
+    GENDER_OPTIONS = ((ContactGender.MALE, 'Male'),
+                      (ContactGender.FEMALE, 'Female'),
+                      (ContactGender.OTHER, 'Other'),
+                      (ContactGender.UNSPECIFIED, 'Unspecified'), )
+
+    registered = models.DateTimeField(auto_now_add=True)
+
+    primary_name = models.CharField(max_length=128)
+    primary_gender = models.PositiveSmallIntegerField(
+        choices=GENDER_OPTIONS, default=ContactGender.UNSPECIFIED)
+    primary_email = models.EmailField()
+    primary_phone = models.CharField(max_length=32)
+    primary_type = models.PositiveSmallIntegerField(
+        choices=CONTACT_TYPE_OPTIONS, default=ContactType.FACULTY)
+    secondary_name = models.CharField(max_length=128, blank=True)
+    secondary_gender = models.PositiveSmallIntegerField(
+        choices=GENDER_OPTIONS, blank=True, default=ContactGender.UNSPECIFIED)
+    secondary_email = models.EmailField(blank=True)
+    secondary_phone = models.CharField(max_length=32, blank=True)
+    secondary_type = models.PositiveSmallIntegerField(
+        choices=CONTACT_TYPE_OPTIONS, blank=True, default=ContactType.FACULTY)
+
+    times_attended = models.PositiveSmallIntegerField(default=0)
+    beginner_delegates = models.PositiveSmallIntegerField()
+    intermediate_delegates = models.PositiveSmallIntegerField()
+    advanced_delegates = models.PositiveSmallIntegerField()
+    spanish_speaking_delegates = models.PositiveSmallIntegerField()
+    chinese_speaking_delegates = models.PositiveSmallIntegerField()
+
+    countrypreferences = models.ManyToManyField(
+        Country, through='CountryPreference')
+    committeepreferences = models.ManyToManyField(
+        Committee, limit_choices_to={'special': True})
+
+    registration_comments = models.TextField(default='', blank=True)
+
+    waitlist = models.BooleanField(default=False)
+    waivers_completed = models.BooleanField(default=False)
+
+    assignments_finalized = models.BooleanField(default=False)
+
+    fees_owed = models.DecimalField(
+        max_digits=6, decimal_places=2, default=Decimal('0.00'))
+    fees_paid = models.DecimalField(
+        max_digits=6, decimal_places=2, default=Decimal('0.00'))
+
+    modified_at = models.DateTimeField(default=timezone.now)
+
+    def balance(self):
+        return self.fees_owed - self.fees_paid
+
+    def update_country_preferences(self, country_ids):
+        '''Given a list of country IDs, first dedupe and filter out 0s, then
+        clear the existing country preferences and construct new ones.'''
+        seen = set()
```

Comment:
I think we can remove the handling for duplicates here if we want to; the frontend has had a feature for awhile to prevent selecting the same country twice.